### PR TITLE
Support creation of draft releases and draft PRs

### LINKED ISSUE #246: Support creation of draft release and draft PRs
The current workflow implies the creation of a Github release before being able to open a PR to opam-repo and therefore before being able to get feedback from its CI.
This means that whenever there's an issue with the release that's uncovered by opam-repo's CI, users have to delete the github release and start all over again.

The main issue here is that for popular tools, some distribution maintainers susbscribe to github releases in order to be notified when a new version comes out so they can package it for their distribution. Deleting releases can then cause trouble.

The prefered solution to this would be to have a way to trigger opam-repo CI builds in order to test a release, ie to test a distribution tarball and its associated opam files but this will require some work on ocaml-ci and will take some time before getting  widely adopted.

In the meantime, it should be fairly easy to add a feature to dune-release allowing the creation of a draft release that we would only undraft once opam-repo's CI goes green and we're sure we can proceed with the actual release.

I suggest the following:
1. Add a `--draft` option to `dune-release publish` so that the release is created as a draft.
2. Add a `--draft` option to `dune-release opam submit` so that the PR is opened as a draft.
3. Add a `dune-release undraft` (or any better suited name) command that would undraft the GH release, update the opam-repo PR and undraft it.

Note that 2 and 3 are required because the distrib tarball link will change as we undraft the GH release, meaning we don't want the PR to be merged accidentally untill we have updated the opam file's `url.src` field.

### PR DESCRIPTION
**Pull request by gpetiot**
~#262 should be merged before~ done

Will fix #246 

- [x] Add a `--draft` option to `dune-release publish` so that the release is created as a draft.
- [x] Add a `--draft` option to `dune-release opam submit` so that the PR is opened as a draft.
- [x] Add a `dune-release undraft` command:
  - [x] writing the release id in a file upon creation so that it can be retrieved while undrafting the release
  - [x] undraft the release on github
  - [x] update the opam file with the new release url
  - [x] writing the pr number in a file upon creation so that it can be retrieved while undrafting the PR
  - [x] undraft the opam-repo PR on github
  - [x] check that we don't try to publish a draft release in a non-draft PR

### CODE REVIEW

File: lib/curl.ml [hunk 1]
```diff
@@ -2,47 +2,9 @@ open Bos_setup
...
-            loop (i + 1) (k + 1)
-    in
-    loop 0 0
-
-let create_release ~version ~msg ~user ~repo =
+let create_release ~version ~msg ~user ~repo ~draft =
   let json =
-    strf "{ \"tag_name\" : \"%s\", \"body\" : \"%s\" }"
-      (escape_for_json version) (escape_for_json msg)
+    strf {|{ "tag_name" : %S, "body" : %S, "draft" : %b }|} version msg draft
```
Comment (gpetiot) [inline lib/curl.ml:7]: It seems the escape function is not  needed anymore when using `{| ... |}`

**Comment by NathanReb** [comment]
Hmm it seems the changelog check isn't working as intended :/

**Review by NathanReb — COMMENTED** [review]
I give this a quick spin and I have a few notes:
- When running in draft mode, I think it would be nice if the logs were adjusted e.g. mentioned creating a draft release, opening a draft PR etc...
- The temp files used to store the PR and Release ids are written at the root of the project as `.draft_pr` and `.draft_release`. This triggers a dune-release warning about the repo being dirty and it is not ideal. I think they should instead be written to the `_build` directory and probably in the same way than the `.url` file, i.e. as `_build/<package_name>-<version>.draft_pr` so that those files can be specific to a certain release at least. We'll probably have to improve the robustness with a centralized tmp folder for dune-release but I think that'd be a good first step.
- The undraft didn't work, I haven't taken the time to debug it yet but if you want to take a look, here's the error message:
```
$ dune-release undraft
[-] Undrafting release
dune-release: [ERROR] Could not find "message" from:
  `Assoc ([("url",
            `String ("https://api.github.com/repos/NathanReb/dune-release-testing/releases/34826727"));
            ("assets_url",
             `String ("https://api.github.com/repos/NathanReb/dune-release-testing/releases/34826727/assets"));
            ("upload_url",
             `String ("https://uploads.github.com/repos/NathanReb/dune-release-testing/releases/34826727/assets{?name,label}"));
            ("html_url",
             `String ("https://github.com/NathanReb/dune-release-testing/releases/tag/v0.2.0%2Btest"));
            ("id", `Int (34826727));
            ("author",
             `Assoc ([("login", `String ("NathanReb")); ("id", `Int (7419360));
                       ("node_id", `String ("MDQ6VXNlcjc0MTkzNjA="));
                       ("avatar_url",
                        `String ("https://avatars2.githubusercontent.com/u/7419360?v=4"));
                       ("gravatar_id", `String (""));
                       ("url",
                        `String ("https://api.github.com/users/NathanReb"));
                       ("html_url", `String ("https://github.com/NathanReb"));
                       ("followers_url",
                        `String ("https://api.github.com/users/NathanReb/followers"));
                       ("following_url",
                        `String ("https://api.github.com/users/NathanReb/following{/other_user}"));
                       ("gists_url",
                        `String ("https://api.github.com/users/NathanReb/gists{/gist_id}"));
                       ("starred_url",
                        `String ("https://api.github.com/users/NathanReb/starred{/owner}{/repo}"));
                       ("subscriptions_url",
                        `String ("https://api.github.com/users/NathanReb/subscriptions"));
                       ("organizations_url",
                        `String ("https://api.github.com/users/NathanReb/orgs"));
                       ("repos_url",
                        `String ("https://api.github.com/users/NathanReb/repos"));
                       ("events_url",
                        `String ("https://api.github.com/users/NathanReb/events{/privacy}"));
                       ("received_events_url",
                        `String ("https://api.github.com/users/NathanReb/received_events"));
                       ("type", `String ("User"));
                       ("site_admin", `Bool (false))]));
            ("node_id", `String ("MDc6UmVsZWFzZTM0ODI2NzI3"));
            ("tag_name", `String ("v0.2.0+test"));
            ("target_commitish", `String ("master")); ("name", `Null);
            ("draft", `Bool (false)); ("prerelease", `Bool (false));
            ("created_at", `String ("2020-12-04T13:44:28Z"));
            ("published_at", `String ("2020-12-04T13:57:38Z"));
            ("assets",
             `List ([`Assoc ([("url",
                               `String ("https://api.github.com/repos/NathanReb/dune-release-testing/releases/assets/29148260"));
                               ("id", `Int (29148260));
                               ("node_id",
                                `String ("MDEyOlJlbGVhc2VBc3NldDI5MTQ4MjYw"));
                               ("name", `String ("dummy-v0.2.0.test.tbz"));
                               ("label", `String (""));
                               ("uploader",
                                `Assoc ([("login", `String ("NathanReb"));
                                          ("id", `Int (7419360));
                                          ("node_id",
                                           `String ("MDQ6VXNlcjc0MTkzNjA="));
                                          ("avatar_url",
                                           `String ("https://avatars2.githubusercontent.com/u/7419360?v=4"));
                                          ("gravatar_id", `String (""));
                                          ("url",
                                           `String ("https://api.github.com/users/NathanReb"));
                                          ("html_url",
                                           `String ("https://github.com/NathanReb"));
                                          ("followers_url",
                                           `String ("https://api.github.com/users/NathanReb/followers"));
                                          ("following_url",
                                           `String ("https://api.github.com/users/NathanReb/following{/other_user}"));
                                          ("gists_url",
                                           `String ("https://api.github.com/users/NathanReb/gists{/gist_id}"));
                                          ("starred_url",
                                           `String ("https://api.github.com/users/NathanReb/starred{/owner}{/repo}"));
                                          ("subscriptions_url",
                                           `String ("https://api.github.com/users/NathanReb/subscriptions"));
                                          ("organizations_url",
                                           `String ("https://api.github.com/users/NathanReb/orgs"));
                                          ("repos_url",
                                           `String ("https://api.github.com/users/NathanReb/repos"));
                                          ("events_url",
                                           `String ("https://api.github.com/users/NathanReb/events{/privacy}"));
                                          ("received_events_url",
                                           `String ("https://api.github.com/users/NathanReb/received_events"));
                                          ("type", `String ("User"));
                                          ("site_admin", `Bool (false))]));
                               ("content_type", `String ("application/x-tar"));
                               ("state", `String ("uploaded"));
                               ("size", `Int (1117));
                               ("download_count", `Int (0));
                               ("created_at", `String ("2020-12-04T13:46:17Z"));
                               ("updated_at", `String ("2020-12-04T13:46:17Z"));
                               ("browser_download_url",
                                `String ("https://github.com/NathanReb/dune-release-testing/releases/download/v0.2.0%2Btest/dummy-v0.2.0.test.tbz"))
                               ])
                      ]));
            ("tarball_url",
             `String ("https://api.github.com/repos/NathanReb/dune-release-testing/tarball/v0.2.0+test"));
            ("zipball_url",
             `String ("https://api.github.com/repos/NathanReb/dune-release-testing/zipball/v0.2.0+test"));
            ("body", `String ("CHANGES:\n\n- Extract library\n"))])
```

**Comment by gpetiot** [comment]
So apparently when updating the release online the `browser_download_url` field is not returned, we would only get one after uploading a release, so we would need to remove the old one to reupload the new one but we can't guess the url to put it in the opam file before we resubmit it, or maybe I missed something?

I moved the `.draft_pr`/`.draft_release` to the build directory by the way.

**Comment by NathanReb** [comment]
> So apparently when updating the release online the `browser_download_url` field is not returned, we would only get one after uploading a release, so we would need to remove the old one to reupload the new one but we can't guess the url to put it in the opam file before we resubmit it, or maybe I missed something

The new url is indeed in the response, see this section from the above error:
```
("browser_download_url",
                                `String ("https://github.com/NathanReb/dune-release-testing/releases/download/v0.2.0%2Btest/dummy-v0.2.0.test.tbz"))
```
It appears to be under `.assets.[0].browser_download_url` in the json response. We need a way to detect which asset is the right one but we can use the file name of the asset and we could eventually set the label when uploading it in `publish` so we can use it here to retrieve it!

**Comment by gpetiot** [comment]
I've fixed this issue when undrafting the release (still having a look at the issues rising when undrafting the PR)

**Comment by gpetiot** [comment]
@NathanReb I implemented the requests to undraft pull requests using the github v4 graphQL API

The following workflow now works without any issue: (replace with your own fork of opam-repo)
```sh
$ dune-release distrib --yes
$ dune-release publish --draft --yes
$ dune-release opam pkg
$ dune-release opam submit --opam-repo gpetiot/opam-repository --draft --yes
$ dune-release undraft --opam-repo gpetiot/opam-repository --yes
```

**Comment by gpetiot** [comment]
I've rebased the branch, made sure we use the sanitized tag for the branch name, and added a hint to use --draft when opening a pr for a draft release.

**Comment by gpetiot** [comment]
I've rebased it again, let me know if it would be better to split this PR into smaller bits, and especially the v3/v4 layer and all the tests.

**Comment by gpetiot** [comment]
@NathanReb @pitag-ha let me know when we can merge this one!

File: tests/bin/draft/dune [hunk 2]
```diff
@@ -0,0 +1,3 @@
+(mdx
```
Comment (NathanReb) [inline tests/bin/draft/dune:1]: Could use cram tests for new tests?

File: tests/bin/draft/run.t [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    [-] Extracting tag from first entry in CHANGES.md
+    [-] Using tag "0.1.0"
+    [+] Tagged HEAD with version 0.1.0
+
+We do the whole dune-release process
+
+(1) distrib
+
+    $ dune-release distrib --dry-run
+    [-] Building source archive
```
Comment (NathanReb) [inline tests/bin/draft/run.t:49]: The output of this command can be ignored, we just want to create the tarball for the next command right?

File: bin/cli.ml [hunk 4]
```diff
@@ -148,6 +148,13 @@ let dry_run =
   in
   named (fun x -> `Dry_run x) Arg.(value & flag & info [ "dry-run" ] ~doc)
 
+let draft =
+  let doc =
+    "Produce a draft release that cannot be merged accidentally and has to be \
```
Comment (NathanReb) [inline bin/cli.ml:153]: How about we flag this as experimental so people don't expect the behaviour to be too stable for now?

**Review by NathanReb — COMMENTED** [review]
It looks great! Thanks!

I have a few comments regarding the `undraft` implementation. It is mostly one big function and it's pretty hard to follow what's in there. Could you try to extract named functions to clarify what are the actions we're taking?
I also noted a lot of code in there is duplicated from other modules so it might make sense that they share the code instead.
Finally I wonder if the PR update code could be simplified by simply fetching the PR branch, updating the opam files there to properly set `url.src`. This would probably make the code simpler and also improve the feature as it would benefit from local changes to the PR, e.g. when the user updated the depends field to get a green CI. If that's too much work we can do it in a separate PR, after releasing the draft mode as it is!

File: bin/cli.ml (hunk 4 above)
Comment (gpetiot) [reply bin/cli.ml:153]: done!

File: tests/bin/draft/dune (hunk 2 above)
Comment (gpetiot) [reply tests/bin/draft/dune:1]: If I remember correctly if we enable the new cram feature we have to upgrade all existing tests as we can't say for which .t file it should be enabled or disabled.

File: tests/bin/draft/run.t (hunk 3 above)
Comment (gpetiot) [reply tests/bin/draft/run.t:49]: Indeed, done!

File: bin/undraft.ml [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+
+let update_opam_file ~dry_run ~url pkg =
+  get_pkg_dir pkg >>= fun dir ->
+  Pkg.opam pkg >>= fun opam_f ->
+  OS.Dir.create dir >>= fun _ ->
+  let dest_opam_file = Fpath.(dir / "opam") in
+  let url = OpamUrl.parse url in
+  Pkg.distrib_file ~dry_run pkg >>= fun distrib_file ->
+  let file = Fpath.to_string distrib_file in
+  let hash algo = OpamHash.compute ~kind:algo file in
```
[the remark this answers is not in the document]
Comment (gpetiot) [reply bin/undraft.ml:20]: It is not necessary indeed, done!

File: bin/undraft.ml [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+        Fpath.pp src name)
+    >>= fun () ->
+    OS.Dir.create ~path:true dst >>= fun _ ->
+    cp "opam" >>= fun () ->
+    cp "url" >>= fun () ->
+    cp "descr" >>= fun () ->
+    (* git add *)
+    Vcs.run_git_quiet vcs ~dry_run ~force:true Cmd.(v "add" % p dst)
+  in
+  let rec prepare_packages = function
```
[the remark this answers is not in the document]
Comment (gpetiot) [reply bin/undraft.ml:136]: Indeed, done!

File: bin/undraft.ml [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+      l "Preparing pull request #%s to %a" pr_id pp_opam_repo opam_repo);
+  let branch = Fmt.strf "release-%s-%s" pkg_name tag in
+  Vcs.get () >>= fun vcs ->
+  OS.Dir.current () >>= fun cwd ->
+  let prepare_package name =
+    (* copy opam, descr and url files *)
+    let dir = name ^ "." ^ version in
+    let src = Fpath.(cwd / "_build" / dir) in
+    let dst = Fpath.(v "packages" / name / dir) in
+    let cp f =
```
[the remark this answers is not in the document]
Comment (gpetiot) [reply bin/undraft.ml:115]: I just realized there was already a Sos.cp function, done!

File: bin/undraft.ml [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+  App_log.status (fun l ->
+      l "Undrafting pull request of package %a %a with id %s" Text.Pp.name
+        pkg_name Text.Pp.version version pr_id);
+  update_opam_file ~dry_run ~url pkg >>= fun () ->
+  App_log.status (fun l ->
+      l "Preparing pull request #%s to %a" pr_id pp_opam_repo opam_repo);
+  let branch = Fmt.strf "release-%s-%s" pkg_name tag in
+  Vcs.get () >>= fun vcs ->
+  OS.Dir.current () >>= fun cwd ->
+  let prepare_package name =
```
[the remark this answers is not in the document]
Comment (gpetiot) [reply bin/undraft.ml:110]: done!

File: tests/bin/draft/dune (hunk 2 above)
Comment (NathanReb) [reply tests/bin/draft/dune:1]: Ah ok, make sense, let's stick to mdx then!

File: bin/undraft.ml [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+  Config.v ~user ~local_repo ~remote_repo [ pkg ] >>= fun config ->
+  (match local_repo with
+  | Some r -> Ok Fpath.(v r)
+  | None -> (
+      match config.local with
+      | Some r -> Ok r
+      | None -> R.error_msg "Unknown local repository."))
+  >>= fun local_repo ->
+  (match remote_repo with
+  | Some r -> Ok r
```
[the remark this answers is not in the document]
Comment (gpetiot) [reply bin/undraft.ml:63]: I tried a few things but the mixing of the result and the option types make it not easier to read. I think it can be improved in a following PR, I would like to do something to what we have in ocamlformat, that is: a Config.t type that is built by overriding the default config with environment, .ocamlformat files and options passed by the user, for dnue-release it can be: current config.t read from the yaml file + command line options, it should be computed once at the beginning of each command and then just passed down each function wihtout having to worry if we have taken each option into account.
What do you think?

(The other way would be to dramatically reduce the command line options of dune-release)

File: bin/opam.ml [hunk 10]
```diff
@@ -178,6 +182,17 @@ let submit ?distrib_uri ~token ~dry_run ~yes ~opam_repo ~user local_repo
   let pkg = List.hd pkgs in
   Pkg.version pkg >>= fun version ->
   Pkg.tag pkg >>= fun tag ->
+  Pkg.build_dir pkg >>= fun build_dir ->
+  Pkg.name pkg >>= fun name ->
+  (if draft then Ok ()
+  else
+    match Config.Draft_release.is_set ~dry_run ~build_dir ~name ~version with
```
[the remark this answers is not in the document]
Comment (gpetiot) [reply bin/opam.ml:189]: I agree, it can be done in a future PR once users start asking for this feature.

**Review by NathanReb — APPROVED** [review]
Let's merge this!
